Create materiallist for collaboration

BacLuc opened the pull request:
- [x] add foreign key from materialList to campCollaboration, and use the username (new)
- [x] create Materiallist when creating a camp ~with name displayname of the creator~ with reference to the created campCollaboration
- [x] create a Materiallist when inviting an ~existing user with the displayname of the invited user~ user, with reference to the created campCollaboration
- [x] ~create a Materiallist when inviting a not existing user with the inviteEmail~ same case as above
- [x] ~rename the Materiallist when the Invitation for a previously not existing user is accepted~
- [x] Zu diskutieren: was machen wir, wenn die Invitation mit einem anderen User akzeptiert wird? -> die MaterialList übernimmt den Namen des referenzierten users
- [x] Falls 2 User den gleichen Displayname haben -> user können trotzdem den namen der liste überschreiben. Dann wird natürlich der username nicht mehr nachgeführt
- [x] Falls die CampCollaboration gelöscht wird, und die CampCollaboration keinen name hatte, wird "NoName" ausgegeben.
-> User können entweder die Liste löschen, oder sie umbenennen.

BacLuc committed:
api: create MaterialList when creating a CampCollaboration

pmattmann commented:
Ich würde eine Frage stellen wollen:
Macht es Sinn die MaterialListe bereits bei der Einladung zu erstellen?

Falls Ja:
 - Müsste die MaterialListe bei einem "Reject" nicht gelöscht werden?

Falls Nein:
  - Die MaterialListe muss beim akzeptieren erstellt werden. 
  - Das wär eher mein Favorit.


Ansonsten sieht es für mich gut aus.

BacLuc commented:
> 



> Ich würde eine Frage stellen wollen: Macht es Sinn die MaterialListe bereits bei der Einladung zu erstellen?
> 
> Falls Ja:
> 
>     * Müsste die MaterialListe bei einem "Reject" nicht gelöscht werden?

Kann man machen, würde ich jetzt nicht, da man je nach dem schon MaterialItems zugewiesen hat.
Das würde ich dem Benutzer überlassen.

> 
> Falls Nein:
> 
>     * Die MaterialListe muss beim akzeptieren erstellt werden.
> 
>     * Das wär eher mein Favorit.
> 
> 
> Ansonsten sieht es für mich gut aus.

Wir haben sonst die Philosophie, dass die Manager schon die Grundstruktur des Lagers aufstellen können, ohne dass die Collaborators akzeptiert haben. (Bei ActivityResponsibles, DayResponsibles).
Ich würde das bei den Materiallists auch durchziehen.
Von dem her würde ich die Materiallist schon beim erstellen erstellen.

carlobeltrame commented on api/src/DataPersister/InvitationDataPersister.php:
```diff
@@ -35,16 +43,31 @@ public function supports($data, array $context = []): bool {
         $campCollaboration = $this->campCollaborationRepository->findByInviteKey($data->inviteKey);
         $campCollaboration->user = $this->security->getUser();
         $campCollaboration->status = CampCollaboration::STATUS_ESTABLISHED;
+        $campCollaboration->tmpInviteEmail = $campCollaboration->inviteEmail;
```
Wieso kannst du diese Property einfach setzen ohne dass sie je irgendwo deklariert wurde?

BacLuc replied:
Because php. Ist ein Hack, ich kann mir auch eine bessere Lösung überlegen, falls gewünscht.

BacLuc replied:
Wird in php 8.2 deprecated, siehe https://wiki.php.net/rfc/deprecate_dynamic_properties + https://github.com/php/php-src/pull/7571

Muss mir was neues überlegen.

BacLuc replied:
gefixt mit https://github.com/ecamp/ecamp3/pull/2452/commits/492458958faa9d20e7eb254bdf5bb05ea59b51d8

carlobeltrame commented on api/src/DataPersister/InvitationDataPersister.php:
```diff
@@ -35,16 +43,31 @@ public function supports($data, array $context = []): bool {
     }
 
+    public function onAfterAccept($data): void {
+        if (null === $data->tmpInviteEmail) {
```
Ist `null ===` wirklich der richtige Check hier? `isset` oder `?? null` nicht nötig?

BacLuc replied:
Momentan wird onBeforeAccept immer davor ausgeführt, von dem her ist das tmpInviteEmail immer gesetzt (je nach dem auf null, aber gesetzt).
Mit php8 wäre wahrscheinlich `?? null`besser, werde ich anpassen.

BacLuc replied:
Gefixt mit https://github.com/ecamp/ecamp3/pull/2452/commits/492458958faa9d20e7eb254bdf5bb05ea59b51d8

carlobeltrame reviewed (approved):
Code und Tests LGTM, merci. Bei den Fragen von @pmattmann könnte ich mit beiden Varianten leben.

Im PeriodDataPersister haben wir erst gerade `updateDaysAndScheduleEntries` eingeführt, was ein eher deklarativer Ansatz wäre im Vergleich zu deinem imperativen Code hier. Stehst du dazu dass der Weg wie du es hier machst in diesem Fall besser ist? Beides wird ja nur selten angepasst, aber bei den Days ist es schon wichtiger dass möglichst immer automatisch ein gültiger Zustand da ist, im Gegensatz zu MaterialLists die der User notfalls auch noch manuell nachbessern kann.

BacLuc commented:
> Code und Tests LGTM, merci. Bei den Fragen von @pmattmann könnte ich mit beiden Varianten leben.
> 
> Im PeriodDataPersister haben wir erst gerade `updateDaysAndScheduleEntries` eingeführt, was ein eher deklarativer Ansatz wäre im Vergleich zu deinem imperativen Code hier. Stehst du dazu dass der Weg wie du es hier machst in diesem Fall besser ist? Beides wird ja nur selten angepasst, aber bei den Days ist es schon wichtiger dass möglichst immer automatisch ein gültiger Zustand da ist, im Gegensatz zu MaterialLists die der User notfalls auch noch manuell nachbessern kann.

Du meinst dass ich eine Methode "updateMaterialList" erstelle, die schaut dass eine Materiallist mit dem richtigen Namen da ist?
Und diese vom CampCollaborationDataPersister und vom InvitationDataPersister aus aufrufe?
Wenn dann hätte ich an https://api-platform.com/docs/core/events/#custom-event-listeners gedacht, da das MaterialList erstellen/updaten nicht unbedingt an den CampDataPersister, den CampCollaborationDataPersister oder den InvitationDataPersister gehängt werden muss, sondern an das erstellen/updaten einer CampCollaboration, egal wer das macht.
Ist die Frage, ob diese Listener das auch können, wenn man einfach selber mit dem EntityManager auf der Datenbank rumgurkt.
Das war mir dann doch bisschen overkill. Aber kann man natürlich machen.

BacLuc commented:
Geil, github lässt die tests nach dem mergen laufen: https://github.com/ecamp/ecamp3/runs/5279881172?check_suite_focus=true

BacLuc commented:
> > Code und Tests LGTM, merci. Bei den Fragen von @pmattmann könnte ich mit beiden Varianten leben.
> > Im PeriodDataPersister haben wir erst gerade `updateDaysAndScheduleEntries` eingeführt, was ein eher deklarativer Ansatz wäre im Vergleich zu deinem imperativen Code hier. Stehst du dazu dass der Weg wie du es hier machst in diesem Fall besser ist? Beides wird ja nur selten angepasst, aber bei den Days ist es schon wichtiger dass möglichst immer automatisch ein gültiger Zustand da ist, im Gegensatz zu MaterialLists die der User notfalls auch noch manuell nachbessern kann.
> 
> Du meinst dass ich eine Methode "updateMaterialList" erstelle, die schaut dass eine Materiallist mit dem richtigen Namen da ist? Und diese vom CampCollaborationDataPersister und vom InvitationDataPersister aus aufrufe? Wenn dann hätte ich an https://api-platform.com/docs/core/events/#custom-event-listeners gedacht, da das MaterialList erstellen/updaten nicht unbedingt an den CampDataPersister, den CampCollaborationDataPersister oder den InvitationDataPersister gehängt werden muss, sondern an das erstellen/updaten einer CampCollaboration, egal wer das macht. Ist die Frage, ob diese Listener das auch können, wenn man einfach selber mit dem EntityManager auf der Datenbank rumgurkt. Das war mir dann doch bisschen overkill. Aber kann man natürlich machen.

Angepasst mit https://github.com/ecamp/ecamp3/pull/2452/commits/6a31b89aaf73d2cc601d5d2858ee16f7c8e7f0d5 und https://github.com/ecamp/ecamp3/pull/2452/commits/371faa65f700f38ba2a992bc4cc9afca3ce9e190.
Dadurch dass man jetzt den update und den create concern zusammenlegt, wird es meiner meinung nach nicht unbedingt besser. Aber ihr dürft auch noch eure Meinung dazu geben.

carlobeltrame commented on api/src/DataPersister/Util/CustomActionListener.php:
```diff
@@ -53,9 +53,9 @@ public function getAfterAction(): Closure {
         $beforeActionReflectionFunction = new ReflectionFunction($beforeAction);
 
-        return 1 != $beforeActionReflectionFunction->getNumberOfParameters();
+        return 2 != $beforeActionReflectionFunction->getNumberOfParameters();
```
Just noticed, the method name describes the opposite of the implementation. How about you change `!=` to `==` here and use it like `!self::hasTwoParameters($beforeAction)`?

carlobeltrame commented on api/src/DataPersister/Util/CustomActionListener.php:
```diff
@@ -22,17 +22,17 @@ public static function of(
-        if (self::hasOneParameter($beforeAction)) {
-            throw new \InvalidArgumentException('beforeAction must have exactly one parameter');
+        if (self::hasTwoParameters($beforeAction)) {
+            throw new \InvalidArgumentException('beforeAction must have exactly two parameter');
```
```suggestion
            throw new \InvalidArgumentException('beforeAction must have exactly two parameters');
```

carlobeltrame commented on api/src/DataPersister/Util/CustomActionListener.php:
```diff
@@ -22,17 +22,17 @@ public static function of(
-        if (self::hasOneParameter($afterAction)) {
-            throw new \InvalidArgumentException('afterAction must have exactly one parameter');
+        if (self::hasTwoParameters($afterAction)) {
+            throw new \InvalidArgumentException('afterAction must have exactly two parameter');
```
```suggestion
            throw new \InvalidArgumentException('afterAction must have exactly two parameters');
```

carlobeltrame commented on api/src/DataPersister/CampCollaborationDataPersister.php:
```diff
@@ -105,4 +123,43 @@ public function onResendInvitation(CampCollaboration $data) {
+    /**
+     * Returns a MaterialList if the MaterialList needs to be updated/created.
+     */
+    public static function getOrCreateUpdatedMaterialListForCollaboration(
```
Wie wäre es wenn du diese Methode in eine Service-Klasse auslagern würdest, die du dann via DI im CampCollaborationDataPersister, CampDataPersister und InvitationDataPersister verwenden kannst? Dann könntest du dir MaterialListRepository dort in der Service-Klasse injecten lassen, was in meinen Augen zu besserer Separation of Concerns führen würde.
In dieser Service-Klasse wäre es dann meiner Meinung nach berechtigt wenn du das `persist` und `flush` auch gleich dort erledigst, statt das in allen drei DataPersisters dupliziert zu machen.
Und wenn diese Methode auch das Datenbank-Persistieren übernimmt könnte der Methodenname stattdessen `updateOrCreateMaterialListForCollaboration` o.ä. sein.

carlobeltrame commented on api/src/DataPersister/CampDataPersister.php:
```diff
@@ -55,6 +58,15 @@ public function afterCreate($data): void {
+
+        $materialList = CampCollaborationDataPersister::getOrCreateUpdatedMaterialListForCollaboration(
+            $this->materialListRepository,
+            dataBefore: null,
```
Dieses Argument könnte auch als Default-Wert `null` haben oder? Es ist ja tatsächlich ein optionales Argument. Findest du es gewinnt an Lesbarkeit wenn man explizit `dataBefore: null` angeben muss? Die named Arguments finde ich hier auch elegant, aber beim Lesen des Methodenaufrufs weiss ich persönlich trotzdem nicht automatisch, was in der Methode mit `dataBefore` geschieht.

Ausserdem würde ich vorschlagen, dieses Argument auf `invitation` o.ä. umzubenennen. Im InvitationDataPersister ist es noch halbwegs klar dass `$dataBefore` eine `Invitation` ist, aber bei der Verwendung im `CampDataPersister` und `CampCollaborationDataPersister` gibt der Argumentname `dataBefore` keine Intuition für den Datentyp der dort reingegeben werden muss.

carlobeltrame commented on api/src/DataPersister/Util/DataPersisterObservable.php:
```diff
@@ -118,11 +118,11 @@ public function persist($data, array $context = []): object {
             call_user_func($this->onAfterCreate, $data);
         } elseif ('patch' === ($context['item_operation_name'] ?? null)) {
-            call_user_func($this->onAfterUpdate, $data);
+            call_user_func($this->onAfterUpdate, $dataBefore, $data);
```
Könnte ein before-Listener die `$dataBefore` mutieren und dann hätten die after-Listener (und andere Orte im Code die `attributes->get('previous_data')` verwenden) nur noch die mutierten Daten zu Verfügung? Müssten wir `$dataBefore` jeweils klonen bevor wir es in die Listeners reingeben? Oder denkst du das ist kein Problem?

carlobeltrame commented on api/src/DataPersister/CampCollaborationDataPersister.php:
```diff
@@ -105,4 +123,43 @@ public function onResendInvitation(CampCollaboration $data) {
+        ?Invitation $dataBefore,
+        CampCollaboration $campCollaboration
+    ): ?MaterialList {
+        if (null !== $dataBefore) {
+            $userDisplayName = $campCollaboration->user?->getDisplayName();
+            $materialList =
+                $userDisplayName ?
+                    $materialListRepository->findByCampAndName($campCollaboration->camp, $userDisplayName) : null;
+            if (null === $materialList) {
+                $materialList = $dataBefore->inviteEmail ?
+                    $materialListRepository->findByCampAndName($campCollaboration->camp, $dataBefore->inviteEmail) : null;
+            }
+            if (null === $materialList) {
+                return null;
+            }
+        } else {
+            $materialList = new MaterialList();
+            $campCollaboration->camp->addMaterialList($materialList);
+        }
+        $newMaterialListName = $campCollaboration->user?->getDisplayName() ??
+            $campCollaboration->inviteEmail ??
+            $dataBefore?->inviteEmail;
+        $existingMaterialList = null != $newMaterialListName ? $materialListRepository->findByCampAndName(
+            $campCollaboration->camp,
+            $newMaterialListName
+        ) : null;
+        if (null !== $existingMaterialList) {
+            return null;
+        }
+        $materialList->name = $newMaterialListName;
+
+        return $materialList;
```
Um diese Methode lesbarer zu machen würde ich vorschlagen, verschiedene Teile der Logik in eigene Methoden auszulagern. Wie wäre es z.B. ungefähr so:
```php
        $materialList = null;
        
        if (null !== $invitation) {
            $materialList = $this->findExistingMaterialList($campCollaboration, $invitation);
        } else {
            $materialList = new MaterialList();
            $campCollaboration->camp->addMaterialList($materialList);
        }
        if (null === $materialList) {
            return null;
        }
        
        $newName = $this->materialListNameFor($campCollaboration, $invitation);
        if ($this->materialListWithSameNameAlreadyExists($campCollaboration->camp, $newName)) {
            return null;
        }
        
        $materialList->name = $newName;
        return $materialList;
```

carlobeltrame replied:
Jetzt da ich das vorherige Refactoring vor mir habe, scheint mir man könnte das Ganze auch noch weiter refactoren, mit einem early exit weniger. Ich bin noch nicht 100% sicher dass ich alle Edge Cases durchgedacht habe, was denkst du, wäre das hier äquivalent?

```php
        $materialList = new MaterialList();
        if (null !== $invitation) {
            $materialList = $this->findExistingMaterialList($campCollaboration, $invitation);
            if (null === $materialList) {
                return null;
            }
        }
        $materialList->name = $this->materialListNameFor($campCollaboration, $invitation);

        // TODO check if the safeguard in addMaterialList works as we would expect, with different object identities but same id
        $campCollaboration->camp->addMaterialList($materialList);
        
        return $materialList;
```

Es ist für mich ohnehin fraglich, ob das eine verbleibende early exit überhaupt wichtig ist. Es deckt ja den Fall ab, dass jemand eingeladen wurde, dann die automatisch erstellte Materialliste manuell gelöscht oder umbenannt wurde (auf etwas anderes als inviteEmail oder displayName), und dann die Einladung akzeptiert wurde. Dank dem early exit würde in diesem Fall keine neue Materialliste entstehen. Aber mir wäre es glaube ich aktuell auch egal, wenn wieder eine entstünde. Wenn wir das akzeptieren würden, könnte man es noch weiter vereinfachen:

```php
        $materialList = $this->findExistingMaterialList($campCollaboration, $invitation) ?? new MaterialList();
        $materialList->name = $this->materialListNameFor($campCollaboration, $invitation);
        $campCollaboration->camp->addMaterialList($materialList);
        return $materialList;
        // Oder gleich $this->em->persist($materialList); $this->em->flush();
```

Die Methode `findExistingMaterialList` würde sich übrigens auch recht simpel gestalten:
```php
    public function findExistingMaterialList (CampCollaboration $campCollaboration, ?Invitation $invitation = null) {
        return $materialListRepository->findByCampAndName(
            $campCollaboration->camp,
            $campCollaboration->user?->getDisplayName() ?? $invitation->inviteEmail
            // Oder falls auch $campCollaboration->inviteEmail okay wäre: $this->materialListNameFor($campCollaboration, $invitation)
        );
    }
```

carlobeltrame reviewed (changes requested):
Kommt gut. Die momentan noch komplexe Methode könnte noch von etwas [XP](http://www.extremeprogramming.org/rules/refactor.html) profitieren ;)

BacLuc committed:
MaterialList: add foreign key to CampCollaboration

That the name of the material list changes with the user displayname and
the status of the campCollaboration (switch from inviteEmail to existing User).

BacLuc commented:
Implementation ist jetzt mit foreign key vereinfacht.

carlobeltrame commented on api/src/Entity/MaterialList.php:
```diff
@@ -109,6 +120,16 @@ public function removeMaterialItem(MaterialItem $materialItem): self {
+        return $this->name
+            ?? $this->campCollaboration?->user?->getDisplayName()
+            ?? $this->campCollaboration?->inviteEmail
+            ?? 'NoName';
```
Ich glaube ich fände es dennoch besser wenn wir einfach `null` ausgeben würden, wenn weder campCollaboration noch Name definiert sind. Im Frontend finde ich folgende Orte wo man einen Fallback auf einen übersetzten String (z.B. "Unbenannte Materialliste") einbauen müsste:
* https://github.com/ecamp/ecamp3/blob/devel/frontend/src/components/campAdmin/CampMaterialListsItem.vue#L4
* https://github.com/ecamp/ecamp3/blob/devel/frontend/src/components/campAdmin/CampMaterialListsItem.vue#L33
* https://github.com/ecamp/ecamp3/blob/devel/frontend/src/components/campAdmin/DialogMaterialListEdit.vue#L7
* https://github.com/ecamp/ecamp3/blob/devel/frontend/src/components/material/MaterialTable.vue#L284
* https://github.com/ecamp/ecamp3/blob/devel/frontend/src/components/print/print-react/components/scheduleEntry/contentNode/Material.jsx#L33
* https://github.com/ecamp/ecamp3/blob/devel/print/components/scheduleEntry/contentNode/Material.vue#L9

Die Übersetzung müsste in den common Übersetzungen landen, da sie auch im Nuxt Print gebraucht wird.

Von mir aus kann der PR hier aber auch gemerged werden, und das könnte in einem good first issue später noch erledigt werden.

usu replied:
Ich würde eher dazu tendieren, dass der return-value von getName nie null sein darf. Könnten wir ja über eine Validation erreichen, dass `name` und `campCollaboration` nicht beide null sein dürfen. Für mich auch ok als good first issue.

carlobeltrame replied:
Auch gut, aber dann müssen wir beim kompletten Austritt einer CampCollaboration den Namen der Liste auf den Namen des ausgetretenen Users setzen

BacLuc replied:
Dann würden wir das mal als KISS so mergen, und einen Issue machen mit:
getName wird nicht null (Validation), + namen setzen, wenn CampCollaboration gelöscht wird?

carlobeltrame replied:
Gut für mich :+1:

BacLuc replied:
In #2574

carlobeltrame reviewed (approved):
Sehr cool wie sich das ganze jetzt vereinfacht hat mit der DB-Relation!

usu commented on api/src/DataPersister/CampCollaborationDataPersister.php:
```diff
@@ -60,12 +63,20 @@ public function beforeCreate($data): BaseEntity {
+        $data->camp->addMaterialList($materialList);
+        $this->em->persist($materialList);
+
+        $this->em->flush();
```
Das flush braucht es, weil wir hier im afterCreate sind?

Würde es Sinn machen, solche Logik im beforeCreate zu haben, und im afterCreate nur Nebeneffekte wie z.B. E-Mails versenden?

BacLuc replied:
Jup, ist analog zu CampDataPersister gemacht.
Ich hätte jetzt side effects eher im afterCreate gesehen. Falls wir technisch das flush loswerden wollen, sollten wir das an beiden Orten ändern.

BacLuc replied:
In https://github.com/ecamp/ecamp3/issues/2574

pmattmann commented on api/src/Entity/MaterialList.php:
```diff
@@ -117,6 +138,6 @@ public function copyFromPrototype($prototype, $entityMap): void {
 
         $this->materialListPrototypeId = $prototype->getId();
-        $this->name = $prototype->name;
+        $this->name = $prototype->getName();
```
Ich würde hier bei `$prototype->name` bleiben.
Ich glaube hier sollte nicht die Anzeige-Logik in gespeicherte Daten übernommen werden.
(z.B. "NoName")

Ebenfalls als "good first issue" machbar.

carlobeltrame replied:
Bin nicht sicher ob ich hier zustimme.
In Lager-Vorlagen rechne ich nicht damit, dass viele CampCollaboration-Materiallisten vorhanden sind, im Optimalfall sogar keine.
Falls wir in Zukunft mal das "klonen" von normalen Lagern erlauben, fände ich es aber besser, wenn ich z.B. noch sehe, welches Material letztes Jahr von wem genau organisiert wurde, statt dass dann fünf "NoName" Materiallisten im Lager sind. Grund ist, dass ich bei der diesjährigen Durchführung möglicherweise diese Leute anfragen möchte, ob sie vorbereitetes Material noch bei sich haben (z.B. digitale Ausbildungsunterlagen).

BacLuc replied:
In https://github.com/ecamp/ecamp3/issues/2574
